File: src/de/umr/fixcon/graphFunctions/standardFunctions/EdgeCountFunction.java
package de.umr.fixcon.graphFunctions.standardFunctions;

import com.google.common.graph.Graph;
import de.umr.fixcon.graphFunctions.AbstractGraphFunction;
import de.umr.fixcon.graphFunctions.GraphFunction;

import static com.google.common.base.Preconditions.checkArgument;

public class EdgeCountFunction extends AbstractGraphFunction implements GraphFunction {

    @Override
    public boolean isEdgeMonotone() {
        return true;
    }

    @Override
    public double apply(final Graph<Integer> g, final int... parameters) {
        return degreeStream(g).sum() / 2.0;
    }

    @Override
    public double optimum(final int size) {
        checkArgument(size >= 0);
        return size * (size - 1) / 2.0;
    }
}

File: src/de/umr/fixcon/graphFunctions/standardFunctions/EdgeCountFunction.kt
package de.umr.fixcon.graphFunctions.standardFunctions

import de.umr.core.dataStructures.VertexOrderedGraph
import de.umr.core.edgeCount
import de.umr.core.vertexCount
import de.umr.fixcon.graphFunctions.GraphFunction
import org.jgrapht.Graph
import org.jgrapht.graph.DefaultEdge

/**This function counts the number of edges in a given graph.*/
object EdgeCountFunction : GraphFunction {

    /**corresponds to the arithmetic series: subgraph.size + subgraph.size+1 + ... + targetSize-1*/
    override fun <V> completeAdditionBound(subgraph: Graph<V, DefaultEdge>, targetSize: Int, args: List<Int>) =
            (subgraph.vertexCount until targetSize).sum()

    override fun <V> eval(g: Graph<V, DefaultEdge>, args: List<Int>): Int = g.edgeCount

    override fun globalOptimum(graphSize: Int) = (graphSize * (graphSize - 1) / 2)
}